### DISCUSSION TITLE
Is it the C# compiler or the .NET runtime who assigns variables their default values?

### DISCUSSION THREAD

**Discussion by dmedine** [dotnet/roslyn#58767] [Q&A] [answered]
This came up on SO recently. https://stackoverflow.com/questions/70661522/what-initializes-instance-fields-with-their-respective-default-value-when-never?noredirect=1#comment124915653_70661522

I assume it is the compiler that gives variables their default values, however several people claim it is the runtime. I have no idea, but if I were designing this, I would make the compiler responsible for converting (e.g.) `double val;` to `double val = 0.0;`.

**Comment by VolodymyrBS** [comment, marked as answer]
I didn't find any official document that describes it(did not spend much time searching), but you could check this yourself. Look at IL of code for SO question [SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgTARgLAChUwMwAJ44QdlQG9UsydsYEA2HAFiwFkBDMAOwAoqAGAbQF0szAE4BzAM4BKUuRIpyCrAGERWAMYiEWALxY2AUwDuykR0kBuVDMVMA9sFvC1ATzUR9WALb3HLt/q1dA2NGHydXdzNLNHkbPABODgASACIAFQALDwAjYWY2ABMsWwAzLGAs9VUwcSwAWjqiDWEEADoAcX1gFWEAITzCswBfBqFC8srvAv0ILGd9VVKJj2asGvrG5rbO4FDpiABNBeFhhpSLK1ibKkTUzJyBoqWKj28HcP812oaiN98IgIdLqhd5+dz9fIFU51MZFF5eWz7ObHYpleF/D7uL4bX5hMGAnZ7GZHUySEZ1c7RBRDVA0mKYHBwEzCYjWLAAB2EYAAbsxgB4eFgAPrNCGFaJszk8vkedjAYVTYnHCVXCh4bhYHY9MVQ6SquRxHD4YWix5U8h0tkMuWa4GIpWktkGuIwY1CxWHZVsumWlAM8TAYQAVzU8pB/38rNVUt5/PVCrxAJ1KoUMZlazY8vd9s9IhT5AZgsJif8OrMTrZCldCdBSbN3suVewNuL+xJJz1Cmd1zdHvb5rIPqAA===).
You could see that there is no logic that sets the value of fields of Car or Motorcycle objects.
The compiler does not emit any instructions to initialize them. it's done by the runtime

**Reply by dmedine** [reply]
Thanks! This is exactly what I was looking for.
